### PR TITLE
Add option to build compiler with jemalloc

### PR DESCRIPTION
**Pull request by aconsroe-hpe**
We are seeing a compile time reduction of around 18% for both hello world and larger compiles when using jemalloc instead of the system allocator.

* Adds `CHPL_HOST_{MEM,JEMALLOC}`
* Adds `CHPL_TARGET_{MEM,JEMALLOC}` which take their default from `CHPL_{MEM,JEMALLOC}` if not explicitly set. The former takes precedence over the latter

Platform differences as follows - more details in commit message:
* Linux: `CHPL_HOST_MEM=jemalloc` only supported when `CHPL_HOST_JEMALLOC=bundled`
* Darwin: `CHPL_HOST_MEM=jemalloc` only supported when `CHPL_HOST_JEMALLOC=system`
* Cygwin: jemalloc not supported (this was previously true)

Some example timings show a 10-25% reduction in compilation time.

wall time in seconds. best of 2

`chpl examples/hello.chpl`

|        | cstdlib | jemalloc |
|--------|--------:|---------:|
| clang  | 4.08    | 3.56     |
| llvm   | 3.96    | 3.41     |

`chpl test/regex/correctness.chpl`

|        | cstdlib | jemalloc |
|--------|--------:|---------:|
| clang  | 7.76    | 6.67     |
| llvm   | 7.40    | 6.26     |

`chpl --fast test/regex/correctness.chpl`

|        | cstdlib |jemalloc |
|--------|--------:|--------:|
| clang  | 17.51   | 16.60   |
| llvm   | 15.13   | 13.41   |

Arkouda on a Cray CS

|        | cstdlib | jemalloc |
|--------|--------:|---------:|
| gcc    | 563.64  | 503.90   |
| llvm   | 624.70  | 525.94   |

Tested on Linux and Mac

### CODE REVIEW

File: third-party/jemalloc/Makefile [hunk 1]
```diff
@@ -2,7 +2,7 @@ ifndef CHPL_MAKE_HOME
 export CHPL_MAKE_HOME=$(shell pwd)/../..
 endif
 
-CHPL_MAKE_HOST_TARGET = --target
+CHPL_MAKE_HOST_TARGET ?= --target
```
Comment (ronawho) [inline third-party/jemalloc/Makefile:5]: I wonder if instead of having a default we should just error if this isn't set by the calling makefile

File: compiler/Makefile [hunk 2]
```diff
@@ -123,6 +123,19 @@ TARGETS = $(CHPL)
 
 LIBS = -lm
 
+ifeq ($(CHPL_MAKE_HOST_MEM), jemalloc)
+ifeq ($(CHPL_MAKE_PLATFORM), darwin)
+JEMALLOC_LIBS=-L$(shell jemalloc-config --libdir) -ljemalloc $(shell jemalloc-config --libs)
+else
+ifeq ($(CHPL_LIB_PIC), jemalloc)
```
Comment (ronawho) [inline compiler/Makefile:130]: Does lib_pic impact our host builds or just target? (I'm not sure myself)

File: compiler/Makefile [hunk 3]
```diff
@@ -123,6 +123,19 @@ TARGETS = $(CHPL)
...
+JEMALLOC_LIBS=$(JEMALLOC_HOST_LIB_DIR)/libjemalloc_pic.a
+else
+JEMALLOC_LIBS=$(JEMALLOC_HOST_LIB_DIR)/libjemalloc.a
```
Comment (ronawho) [inline compiler/Makefile:133]: What's the motivation for explicitly specifying the `.a` instead of sticking with `-L`/`-l`. We only build the .a for our bundled version anyways, so there's no shared lib to take precedence. Oh, maybe this is concern about finding a system version instead of finding our bundled version?

I'd personally be in favor of `-L`/`-l` and we probably need to use `jemalloc-config` still to get any extra link args (like say `-lm  -lpthread`)

Comment (ronawho) [reply compiler/Makefile:133]: (And it might be nice to put that logic into a chplenv script to mimic how we're getting links args for the target build)

File: util/chplenv/chpl_mem.py [hunk 4]
```diff
@@ -8,18 +8,18 @@
...
-                mem_val = 'cstdlib'
-            else:
-                mem_val = 'jemalloc'
+    platform_val = chpl_platform.get('target')
+    cygwin = platform_val.startswith('cygwin')
+    if flag == 'target':
+        if cygwin:
+            mem_val = 'cstdlib'
+        else:
+            mem_val = overrides.get('CHPL_TARGET_MEM') or overrides.get('CHPL_MEM') or 'cstdlib'
```
Comment (ronawho) [inline util/chplenv/chpl_mem.py:17]: I think this has changed the default target memory layer to be `cstdlib`

**Comment by ronawho** [comment]
There are almost certainly some doc changes required for this. I'd look for existing docs for `CHPL_MEM`/`CHPL_JEMALLOC`

File: util/chplenv/printchplenv.py [hunk 5]
```diff
@@ -91,7 +91,8 @@
     ChapelEnv('CHPL_HOST_MEM', INTERNAL, 'hostmem'),
     ChapelEnv('CHPL_TARGET_MEM', INTERNAL, 'tgtmem'),
     ChapelEnv('CHPL_MEM', RUNTIME | LAUNCHER | DEFAULT, 'mem'),
```
Comment (ronawho) [inline util/chplenv/printchplenv.py:93]: Should this be split into `CHPL_HOST_MEM` and `CHPL_TARGET_MEM`?

File: util/chplenv/chpl_mem.py (hunk 4 above)
Comment (ronawho) [reply util/chplenv/chpl_mem.py:17]: The old style is more verbose and has more indentation, but I personally find that easier to follow compared to the `or` chaining

File: compiler/Makefile (hunk 3 above)
Comment (ronawho) [reply compiler/Makefile:133]: Due to the lack of `-lpthread` this fails for me on raptor with:

```
/lus/cls04005/eronagha/chapel/third-party/jemalloc/jemalloc-src/src/mutex.c:103: undefined reference to `pthread_mutexattr_destroy'
/lus/cls04005/eronagha/chapel/third-party/jemalloc/install/host/cray-cs-x86_64-none-gnu-none/lib/libjemalloc.a(tcache.o): In function `chpl_je_tsd_set':
```

Resolving the undefined reference should be easy, though I thought it was weird the function name was prefixed with `chpl_je` still.

Comment (ronawho) [reply compiler/Makefile:133]: Oh, nevermind, that's from `--with-private-namespace=chpl_`, not because we're still specifying a `--with-jemalloc-prefix`

**Comment by ronawho** [comment]
There's some logic that triggers a compiler rebuild if a component changed. Like if you switch from LLVM=none to bundled, we'll rebuild or something. I _think_ we may need to do that here too? I would guess Michael knows more about this.

**Comment by ronawho** [comment]
FYI here's the time to build on Arkouda on a Cray CS (16-node-cs-hdr) before and after this change:

before:
```
Pass               Name                   Main    Check    Clean     Time    %     Accum    %
---- ---------------------------------  -------  -------  -------  ------- -----  ------- -----
  42 makeBinary                         334.503    0.000    2.436  336.939  58.3  336.939  58.3
  13 resolve                            147.948    0.102    1.838  149.888  25.9  486.827  84.2
  22 parallel                            11.299    0.076    0.266   11.640   2.0  498.467  86.2
  20 callDestructors                     10.083    0.000    0.197   10.280   1.8  508.747  88.0
  29 copyPropagation                      8.645    0.057    0.794    9.497   1.6  518.244  89.6
  41 codegen                              7.710    0.000    0.171    7.881   1.4  526.125  91.0
  36 insertWideReferences                 7.423    0.051    0.206    7.680   1.3  533.805  92.3
  26 inlineFunctions                      5.743    0.122    0.967    6.832   1.2  540.637  93.5
  33 loopInvariantCodeMotion              5.791    0.048    0.121    5.960   1.0  546.596  94.6
  21 lowerIterators                       4.565    0.080    0.652    5.297   0.9  551.894  95.5
  ...
     total time                         563.640    1.264   13.174  578.078
```

after:
```
Pass               Name                   Main    Check    Clean     Time    %     Accum    %
---- ---------------------------------  -------  -------  -------  ------- -----  ------- -----
  42 makeBinary                         332.513    0.000    0.742  333.255  65.2  333.255  65.2
  13 resolve                            113.196    0.071    0.600  113.866  22.3  447.121  87.5
  22 parallel                             7.347    0.052    0.214    7.614   1.5  454.735  89.0
  20 callDestructors                      6.672    0.000    0.122    6.794   1.3  461.529  90.3
  36 insertWideReferences                 6.068    0.037    0.136    6.242   1.2  467.771  91.5
  29 copyPropagation                      5.772    0.042    0.308    6.122   1.2  473.892  92.7
  41 codegen                              5.567    0.000    0.125    5.692   1.1  479.584  93.8
  26 inlineFunctions                      4.705    0.078    0.562    5.346   1.0  484.930  94.9
  33 loopInvariantCodeMotion              4.531    0.033    0.102    4.667   0.9  489.596  95.8
  21 lowerIterators                       3.897    0.054    0.330    4.282   0.8  493.878  96.6
  ...
       total time                         503.901    0.897    6.236  511.035
```

This was with `CHPL_LLVM=none` if that matters. Also I'm wondering if there's any advantage to replacing the memory allocator for the C/llvm backend? For C backend we'd probably need a .so and `LD_LIBRARY_PATH`, but it's not clear to me if the current approach gets us using jemalloc with clang/llvm or if there'd be more to do there?

File: util/chplenv/chpl_jemalloc.py [hunk 6]
```diff
@@ -22,6 +29,17 @@ def get():
     if mem_val != 'jemalloc' and jemalloc_val != 'none':
       error("CHPL_JEMALLOC must be 'none' when CHPL_MEM is not jemalloc")
 
+    if flag == 'host':
+      platform_val = chpl_platform.get('target')
+      linux = platform_val.startswith('linux')
+      darwin = platform_val.startswith('darwin')
+
+      if linux and mem_val == 'jemalloc' and jemalloc_val == 'system':
+        error("CHPL_JEMALLOC=system is not supported on Linux for host builds")
```
Comment (ronawho) [inline util/chplenv/chpl_jemalloc.py:38]: This is probably captured in an issue, but what prevents us from using a system jemalloc on linux (I don't actually think this is an important use-case, more just curious)

File: util/chplenv/chpl_jemalloc.py [hunk 7]
```diff
@@ -22,6 +29,17 @@ def get():
...
+
+      if darwin and mem_val == 'jemalloc' and jemalloc_val == 'bundled':
+        error("CHPL_JEMALLOC=bundled is not supported on Mac for host ubilds")
```
Comment (ronawho) [inline util/chplenv/chpl_jemalloc.py:41]: This is probably captured in an issue, but what prevents us from using the bundled jemalloc on mac?  Is it just that we're only building the static library and we need a dylib on mac for things to work correctly? If so we could specialize the host build for mac to produce the dynamic lib instead of building the static lib only.

File: util/chplenv/chpl_jemalloc.py (hunk 6 above)
Comment (aconsroe-hpe) [reply util/chplenv/chpl_jemalloc.py:38]: That was my call because different distros package it so differently, some provide `jemalloc-config`, some have the `.pc` file, fedora has a `.so.n` with no `.so`.

File: util/chplenv/chpl_jemalloc.py (hunk 7 above)
Comment (aconsroe-hpe) [reply util/chplenv/chpl_jemalloc.py:41]: Nothing preventing us. That was my call after running into some other bundled shared library weirdness (see [this comment](https://github.com/Cray/chapel-private/issues/2629#issuecomment-952324222)). I can add it and it would work as well as the other shared libraries I think. (Also I typo'd `ubilds`)

File: util/chplenv/printchplenv.py (hunk 5 above)
Comment (aconsroe-hpe) [reply util/chplenv/printchplenv.py:93]: Yeah that was the direction that got the most traction so that eg an unsupported combo for the compiler wouldn't prevent you from using `jemalloc` in the runtime.

File: third-party/jemalloc/Makefile (hunk 1 above)
Comment (aconsroe-hpe) [reply third-party/jemalloc/Makefile:5]: Yeah that sounds good. Initially I think I made this change before specifying for the runtime build as well so it shouldn't be necessary at this point

**Comment by aconsroe-hpe** [comment]
>  Also I'm wondering if there's any advantage to replacing the memory allocator for the C/llvm backend?

Yeah that's an interesting question. For gnu, the only way I can think of is `LD_PRELOAD` on Linux and even then, I'm not sure if there's only single process level (ie. can we be sure the env var gets propagated to all descendents?).

For builtin clang, we could build llvm+clang with jemalloc and try that out.

File: util/chplenv/chpl_mem.py (hunk 4 above)
Comment (aconsroe-hpe) [reply util/chplenv/chpl_mem.py:17]: Do you like this better?

```python
@memoize
def get(flag='host'):
    platform_val = chpl_platform.get('target')
    cygwin = platform_val.startswith('cygwin')
    chpl_host_mem = overrides.get('CHPL_HOST_MEM')
    chpl_target_mem = overrides.get('CHPL_TARGET_MEM')
    chpl_mem = overrides.get('CHPL_MEM')

    if flag == 'target':
        if cygwin:
            mem_val = 'cstdlib'
        elif chpl_target_mem:
            mem_val = chpl_target_mem
        elif chpl_mem:
            mem_val = chpl_mem
        else:
            mem_val = 'jemalloc'
    elif flag == 'host':
        if cygwin:
            mem_val = 'cstdlib'
        elif chpl_host_mem:
            mem_val = chpl_host_mem
        elif chpl_mem:
            mem_val = chpl_mem
        else:
            mem_val = 'cstdlib'
    else:
        error("Invalid flag: '{0}'".format(flag), ValueError)
    return mem_val
```

Comment (ronawho) [reply util/chplenv/chpl_mem.py:17]: Yeah, I like that. Definitely more code, but I find it easier to parse. Is the plan to eventually default to jemalloc for the host too, or do we imagine that being opt-in for now?

File: third-party/jemalloc/Makefile (hunk 1 above)
Comment (aconsroe-hpe) [reply third-party/jemalloc/Makefile:5]: That would be ideal, but it looks like its already set with `ifndef` in `make/Makefile.base` so we can't easily error on it not being set. Introduce another jemalloc var that specifically has to be set? Leave as is?

File: compiler/Makefile (hunk 2 above)
Comment (aconsroe-hpe) [reply compiler/Makefile:130]: Yeah I'm only seeing for target builds, will fix

File: third-party/jemalloc/Makefile (hunk 1 above)
Comment (ronawho) [reply third-party/jemalloc/Makefile:5]: Leave as-is

File: compiler/Makefile (hunk 2 above)
Comment (ronawho) [reply compiler/Makefile:130]: I think we only use it for target builds, and I think that's fine. I don't know that we need PIC support for the compile (and if we only support for target, then I think we can remove the PIC linking here)

File: util/chplenv/chpl_mem.py (hunk 4 above)
Comment (aconsroe-hpe) [reply util/chplenv/chpl_mem.py:17]: I think opt-in for now until we get further testing done. Have updated to use the above

File: compiler/Makefile (hunk 3 above)
Comment (aconsroe-hpe) [reply compiler/Makefile]: Updated to use `jemalloc-config` for bundled builds in `chpl_jemalloc.py`. We should do the same for `system` right?

Comment (ronawho) [reply compiler/Makefile]: System is only supported with dynamic libraries today, right? So shouldn't matter since the dynamic libs will figure out their dependencies.

File: util/chplenv/chpl_jemalloc.py [hunk 8]
```diff
@@ -1,26 +1,55 @@
...
+        elif chpl_jemalloc:
+            jemalloc_val = chpl_jemalloc
+        elif linux and mem_val == 'jemalloc':
             jemalloc_val = 'bundled'
+        elif darwin and mem_val == 'jemalloc':
+            jemalloc_val = 'system'
         else:
             jemalloc_val = 'none'
 
+    # TODO what to do on crays?
```
Comment (aconsroe-hpe) [inline util/chplenv/chpl_jemalloc.py:40]: @ronawho forgot to bring this up, is there a preferred way to test for linux-y systems including cray?

Comment (ronawho) [reply util/chplenv/chpl_jemalloc.py:40]: I think we tend to just assume if it's not mac or cygwin, then it's linux-y.

File: util/chplenv/chpl_jemalloc.py [hunk 9]
```diff
@@ -1,26 +1,55 @@
...
 from utils import error, memoize, run_command, warning
 
 
 @memoize
-def get():
-    jemalloc_val = overrides.get('CHPL_JEMALLOC')
-    mem_val = chpl_mem.get('target')
-    if not jemalloc_val:
-        if mem_val == 'jemalloc':
+def get(target='target'):
```
Comment (ronawho) [inline util/chplenv/chpl_jemalloc.py:11]: Using the arg name of `target` here feels a little confusing to me since target=host feels slightly off. In other chplenv scripts I think we just call the name `flag`, which isn't as descriptive but doesn't have the double use of target.

**Comment by aconsroe-hpe** [comment]
> There's some logic that triggers a compiler rebuild if a component changed. Like if you switch from LLVM=none to bundled, we'll rebuild or something. I think we may need to do that here too? I would guess Michael knows more about this.

This is now handled by [this commit](https://github.com/chapel-lang/chapel/pull/18627/commits/91f2f29a391c313ba8ee722f244df95f9cdd7f57)

**Comment by ronawho** [comment]
FYI here's timings for 16-node-cs-hdr with llvm this time instead of the c backend:

before:
```
Pass               Name                   Main    Check    Clean     Time    %     Accum    %
---- ---------------------------------  -------  -------  -------  ------- -----  ------- -----
  42 makeBinary                         350.803    0.000    2.477  353.280  55.2  353.280  55.2
  13 resolve                            157.137    0.110    1.897  159.144  24.9  512.425  80.1
  41 codegen                             39.173    0.000    0.134   39.307   6.1  551.732  86.3
  22 parallel                            12.525    0.080    0.268   12.874   2.0  564.606  88.3
  20 callDestructors                     11.554    0.000    0.220   11.775   1.8  576.380  90.1
  29 copyPropagation                      9.543    0.059    0.801   10.403   1.6  586.783  91.8
  36 insertWideReferences                 6.926    0.053    0.212    7.191   1.1  593.974  92.9
  26 inlineFunctions                      6.062    0.127    0.998    7.187   1.1  601.161  94.0
  33 loopInvariantCodeMotion              6.190    0.048    0.125    6.364   1.0  607.525  95.0
  21 lowerIterators                       4.765    0.083    0.676    5.525   0.9  613.050  95.9
  ...
     total time                         624.703    1.320   13.399  639.423
```

after:
```
Pass               Name                   Main    Check    Clean     Time    %     Accum    %
---- ---------------------------------  -------  -------  -------  ------- -----  ------- -----
  42 makeBinary                         303.885    0.000    0.806  304.691  57.1  304.691  57.1
  13 resolve                            131.109    0.073    0.635  131.816  24.7  436.507  81.8
  41 codegen                             33.943    0.000    0.114   34.057   6.4  470.564  88.2
  22 parallel                             8.701    0.053    0.220    8.973   1.7  479.537  89.9
  20 callDestructors                      6.808    0.000    0.125    6.933   1.3  486.470  91.2
  29 copyPropagation                      6.261    0.042    0.316    6.619   1.2  493.088  92.4
  36 insertWideReferences                 6.012    0.038    0.140    6.191   1.2  499.279  93.6
  33 loopInvariantCodeMotion              5.637    0.038    0.123    5.798   1.1  505.077  94.7
  26 inlineFunctions                      4.946    0.079    0.575    5.599   1.0  510.676  95.7
  21 lowerIterators                       4.041    0.055    0.337    4.432   0.8  515.109  96.6
  ...
     total time                         525.937    0.939    6.635  533.511
```

So more of an improvement since jemalloc is being used for parts of the backend too.

File: util/chplenv/third_party_utils.py [hunk 10]
```diff
@@ -25,8 +25,13 @@ def default_uniq_cfg_path():
...
 @memoize
-def get_cfg_install_path(pkg, ucp=default_uniq_cfg_path()):
-    return os.path.join(get_chpl_third_party(), pkg, 'install', ucp)
+def get_cfg_install_path(pkg, ucp=default_uniq_cfg_path(), host_or_target=None):
+    if host_or_target is None:
+        return os.path.join(get_chpl_third_party(), pkg, 'install', ucp)
+    else:
+        # some third party packages (like jemalloc) can be built for both host and target
+        # in which case they are under jemalloc/install/{host,target}/... for example
+        return os.path.join(get_chpl_third_party(), pkg, 'install', host_or_target, ucp)
```
Comment (ronawho) [inline util/chplenv/third_party_utils.py:34]: Seems like this could be slightly simpler if `host_or_target` defaulted to `''` instead of `None` and we just always did the join with that?

File: util/chplenv/printchplenv.py [hunk 11]
```diff
@@ -258,20 +263,22 @@ def compute_internal_values():
...
 
     ENV_VALS['  CHPL_PE_CHPL_PKGCONFIG_LIBS'] = chpl_llvm.gather_pe_chpl_pkgconfig_libs()
 
+""" Remove duplicates, keeping last occurrence and preserving order
+e.g. "-lhwloc -lqthread -lhwloc ..." -> "-lqthread -lhwloc ..."""
+def dedup(args):
+    seen = set()
+    ret = [arg for arg in reversed(args)
+           if not (arg in seen or seen.add(arg))]
+    return reversed(ret)
```
Comment (ronawho) [inline util/chplenv/printchplenv.py:280]: How do we end up with duplicates and did this PR change anything with respect to that, or is that just something you noticed while working on this?

Comment (ronawho) [reply util/chplenv/printchplenv.py:280]: Oh, duh -- nevermind. You just moved dedup into a helper function to use it in more places.

File: util/chplenv/printchplenv.py [hunk 12]
```diff
@@ -176,8 +178,10 @@ def compute_all_values():
     ENV_VALS['CHPL_LAUNCHER'] = chpl_launcher.get()
     ENV_VALS['CHPL_TIMERS'] = chpl_timers.get()
     ENV_VALS['CHPL_UNWIND'] = chpl_unwind.get()
+    ENV_VALS['CHPL_HOST_MEM'] = chpl_mem.get('host')
     ENV_VALS['CHPL_MEM'] = chpl_mem.get('target')
-    ENV_VALS['  CHPL_JEMALLOC'] = chpl_jemalloc.get()
+    ENV_VALS['  CHPL_HOST_JEMALLOC'] = chpl_jemalloc.get('host')
```
Comment (ronawho) [inline util/chplenv/printchplenv.py:183]: Today the printed order is:

```
CHPL_HOST_MEM: jemalloc *
CHPL_TARGET_MEM: jemalloc
CHPL_MEM: jemalloc
  CHPL_HOST_JEMALLOC: bundled
  CHPL_TARGET_JEMALLOC: bundled
```

And I wonder if this would be slightly cleaner as:

```
CHPL_MEM: jemalloc
CHPL_HOST_MEM: jemalloc *
  CHPL_HOST_JEMALLOC: bundled
CHPL_TARGET_MEM: jemalloc
  CHPL_TARGET_JEMALLOC: bundled
```

File: util/chplenv/printchplenv.py [hunk 13]
```diff
@@ -88,10 +88,11 @@
     ChapelEnv('CHPL_LAUNCHER', LAUNCHER | DEFAULT, 'launch'),
     ChapelEnv('CHPL_TIMERS', RUNTIME | LAUNCHER | DEFAULT, 'tmr'),
     ChapelEnv('CHPL_UNWIND', RUNTIME | LAUNCHER | DEFAULT, 'unwind'),
-    ChapelEnv('CHPL_HOST_MEM', INTERNAL, 'hostmem'),
-    ChapelEnv('CHPL_TARGET_MEM', INTERNAL, 'tgtmem'),
+    ChapelEnv('CHPL_HOST_MEM', COMPILER, 'hostmem'),
+    ChapelEnv('CHPL_TARGET_MEM', INTERNAL, 'mem'),
```
Comment (ronawho) [inline util/chplenv/printchplenv.py:92]: Why the change from `tgtmem` to `mem`?

File: util/chplenv/chpl_jemalloc.py [hunk 14]
```diff
@@ -1,26 +1,60 @@
...
-    mem_val = chpl_mem.get('target')
-    if not jemalloc_val:
-        if mem_val == 'jemalloc':
+def get(flag='target'):
+    chpl_jemalloc = overrides.get('CHPL_JEMALLOC')
+    chpl_host_jemalloc = overrides.get('CHPL_HOST_JEMALLOC')
+    chpl_target_jemalloc = overrides.get('CHPL_TARGET_JEMALLOC')
+    mem_val = chpl_mem.get(flag)
+
+    platform_val = chpl_platform.get('target')
```
Comment (ronawho) [inline util/chplenv/chpl_jemalloc.py:17]: Should platform_val be based on `flag` instead of hardcoding to `target `?

File: util/chplenv/chpl_jemalloc.py [hunk 15]
```diff
@@ -33,29 +67,39 @@ def get_uniq_cfg_path():
 # Instead of libtool or pkg-config, jemalloc uses a jemalloc-config script to
 # determine dependencies/link args . It's located in the bin directory
 @memoize
-def get_jemalloc_config_file():
-    install_path = third_party_utils.get_cfg_install_path('jemalloc')
+def get_jemalloc_config_file(target='target'):
```
Comment (ronawho) [inline util/chplenv/chpl_jemalloc.py:70]: `target=` -> `flag=`?

File: util/chplenv/chpl_jemalloc.py [hunk 16]
```diff
@@ -33,29 +67,39 @@ def get_uniq_cfg_path():
...
-    if target_mem == 'bundled':
-        jemalloc_config = get_jemalloc_config_file()
+def get_link_args(target, mem_val):
+    if mem_val == 'bundled':
+        jemalloc_config = get_jemalloc_config_file(target)
         libs = ['-ljemalloc']
+        # should this be an error if we can't find it?
         if os.access(jemalloc_config, os.X_OK):
+            jemalloc_libdir = run_command([jemalloc_config, '--libdir']).strip()
+            libs.append(f'-L{jemalloc_libdir}')
```
Comment (ronawho) [inline util/chplenv/chpl_jemalloc.py:84]: Why is `--libdir` needed now? I thought we got our `-L` from other parts of our scripts.

File: compiler/Makefile [hunk 17]
```diff
@@ -158,7 +158,7 @@ $(CHPL_CONFIG_CHECK): | $(CHPL_BIN_DIR)
 	echo $(CHPL_MAKE_COMPILER_SUBDIR) > $(CHPL_CONFIG_CHECK)
 
 $(CHPL): $(CHPL_OBJS) $(CHPL_CONFIG_CHECK) | $(CHPL_BIN_DIR)
-	$(CXX) $(LDFLAGS) -o $@ $(CHPL_OBJS) $(LIBS) $(LLVM_LIBS)
+	$(CXX) $(LDFLAGS) -o $@ $(CHPL_OBJS) $(CHPL_MAKE_THIRD_PARTY_HOST_LINK_ARGS) $(LIBS) $(LLVM_LIBS)
```
Comment (ronawho) [inline compiler/Makefile:161]: Not needed for this PR, but I wonder if we should fold `LLVM_LIBS` into `CHPL_MAKE_THIRD_PARTY_HOST_LINK_ARGS`

**Comment by ronawho** [comment]
Left a couple of inline comments, but overall I think this is in good shape. Can you add some perf results (maybe in table form) to the PR description to motivate the change?

I think the total time from 16-node-cs-hdr arkouda results and your numbers from https://github.com/Cray/chapel-private/issues/2558#issuecomment-970640581 would be great.

File: util/chplenv/printchplenv.py (hunk 13 above)
Comment (aconsroe-hpe) [reply util/chplenv/printchplenv.py:92]: My thought was that since the intention for `CHPL_TARGET_MEM` to replace `CHPL_MEM`, they should have the same name. I don't think this has an effect either way today

File: util/chplenv/chpl_jemalloc.py (hunk 14 above)
Comment (aconsroe-hpe) [reply util/chplenv/chpl_jemalloc.py:17]: yes good catch

File: util/chplenv/third_party_utils.py (hunk 10 above)
Comment (aconsroe-hpe) [reply util/chplenv/third_party_utils.py:34]: TIL os.path.join removes empty string. I'm happy with that approach. Thanks for the suggestion

File: util/chplenv/chpl_jemalloc.py (hunk 16 above)
Comment (aconsroe-hpe) [reply util/chplenv/chpl_jemalloc.py:84]: I think I threw it in there thinking it was needed when using a system install, if eg the directory is not on the lib path. But it does build without it on Mac where this would come into play

File: util/chplenv/printchplenv.py (hunk 12 above)
Comment (aconsroe-hpe) [reply util/chplenv/printchplenv.py:183]: How do you like 
```
CHPL_HOST_MEM: jemalloc *
  CHPL_HOST_JEMALLOC: bundled
CHPL_MEM: jemalloc
CHPL_TARGET_MEM: jemalloc
  CHPL_TARGET_JEMALLOC: bundled
```
in order to keep `CHPL_MEM` and `CHPL_TARGET_MEM` next to each other?

File: util/chplenv/chpl_mem.py [hunk 18]
```diff
@@ -3,23 +3,35 @@
...
-        mem_val = overrides.get('CHPL_MEM')
-        if not mem_val:
-            platform_val = chpl_platform.get('target')
-            cygwin = platform_val.startswith('cygwin')
-
-            if cygwin:
-                mem_val = 'cstdlib'
-            else:
-                mem_val = 'jemalloc'
+    platform_val = chpl_platform.get('target')
```
Comment (ronawho) [inline util/chplenv/chpl_mem.py:10]: `'target'` -> `flag` here too?

Comment (aconsroe-hpe) [reply util/chplenv/chpl_mem.py:10]: Yes good catch. I grepped `chpl_mem` and `chpl_jemalloc` for `.get('target')` and I think that was the last remaining place.

**Review by ronawho — APPROVED** [review]

File: util/chplenv/chpl_jemalloc.py (hunk 16 above)
Comment (aconsroe-hpe) [reply util/chplenv/chpl_jemalloc.py]: I remember now! I did need that to build bundled. The last commit for this PR is a bit nonsensical actually.
There isn't a bunch of precedent for link args for `chpl`, only llvm, which sets its link args in the Makefile calling into `llvm-config`. I'm inclined to leave this in as it is, especially if we like the idea of `CHPL_MAKE_THIRD_PARTY_HOST_LINK_ARGS`. The counterpart `CHPL_MAKE_THIRD_PARTY_LINK_ARGS` does also include some `-L` for eg `hwloc`.

Comment (aconsroe-hpe) [reply util/chplenv/chpl_jemalloc.py]: And thinking about this again, it seems a bit unnecessary to call `jemalloc-config --libdir` when we could easily construct the right lib path with `os.path.join(install_path, 'lib')`. Does that sound good?

Comment (ronawho) [reply util/chplenv/chpl_jemalloc.py]: Ah, makes sense. Yeah, doing `os.path.join(install_path, 'lib')` would match how we get the `-L` for other packages in `default_get_link_args`

Comment (aconsroe-hpe) [reply util/chplenv/chpl_jemalloc.py]: Fixed in https://github.com/chapel-lang/chapel/pull/18627/files#diff-c6c2f75f0251acf7b7553a7c96e1fddd18db10e53554259d05c990c3229b6a10R78-R84

**Comment by aconsroe-hpe** [comment]
I've also added `test-host-jemalloc.bash` so we can add to nightly testing

**Review by ronawho — APPROVED** [review]
